src/core/components/files-regular/cat-pull-stream.js

```diff
@@ -14,26 +14,40 @@ module.exports = function (self) {
     options = options || {}
 
     ipfsPath = normalizePath(ipfsPath)
+
     const pathComponents = ipfsPath.split('/')
-    const fileNameOrHash = pathComponents[pathComponents.length - 1]
+    const cid = pathComponents[0]
 
     if (options.preload !== false) {
       self._preload(pathComponents[0])
     }
 
     const d = deferred.source()
 
+    let closestMatchedLink
+
     pull(
-      exporter(ipfsPath, self._ipld, options),
-      pull.filter(file => file.path === fileNameOrHash),
+      exporter(cid, self._ipld, options),
```

Comment:
If you export the CID not CID+path you'll have to traverse through the entire graph until you get to the path you were looking for. This could make `cat` VERY slow for big directories if the content is not in your repo already (and be slow anyway, even if it is).

cc @achingbrain

Reply:
+1 that - if you pass the path as well as the CID (e.g. `/ipfs/Qmfoo/path/to/file.txt`) the the unixfs exporter is smart enough to not load whole DAGs for sharded directories, etc which saves a load of time & bandwidth.  Maybe this change is better made there?